docs/settings.md
# Settings

Settings are managed with environment variables. You can set the enviroment variables or use the `.env` file. The `.env` file must be placed in the root folder.

!!! note
    The environment variables and the `.env` file don't have to be all upercase. It's case insensitive.

Here is a list of all the settings clasified by category.

!!! note
    Required settings are marked with 🚩.

## General

- `production`: if not present or `False`, the server will be running in debug mode. If it's True the documentation will be deactivated. Defaults to `False`.

## Security

- `encryption_algorithm`: algorithm use to hash passwords. Don't change this unless you are 100% sure. Default is `HS256`.
- `server_secret`: 32-bytes base64 encoded token used to encrypt and decrypt JSON Web Tokens. To get a valid secret using python, execute `python -c "import secrets;print(secrets.token_urlsafe(32))"`. It's defined in runtime, but it's possible that for each request the token is redifined, so all JSON Web Tokens will be invalid. Therefore, it's very encouraged to define it.
- `token_expire_minutes`: number of minutes before the JSON Web Token expires. Defaults to 30 minutes.

## Database

- 🚩 `first_superuser_password`: first admin password. See the [database setup](database.md#first-admin-settings) for more info.
- 🚩 `first_superuser`: first admin username. See the [database setup](database.md#first-admin-settings) for more info.
- 🚩 `sqlalchemy_database_url`: path of the database. For sql must be like `mysql+pymysql://<user>:<password>@<host>:<port>/<table>`

## Dialogflow

- 🚩 `dialogflow_project_id`: dialogflow's project id.
- 🚩 `google_application_credentials`: path to the google credential's json.

## Logging

- 🚩 `log_path`: absolute path to the log file. The folder and the folder's parents are created at runtime.
- `logging_level`: logging level. Must be `DEBUG`, `INFO`, `WARNING`, `ERROR` or `CRITICAL`. Defaults to `INFO`.

## Bot generics

- `bot_split_char`: string to separate messages of the bot besides `\n`. Defaults to `~`.

!!! example "settings.bot_split_char"
    If dialogflow's chatbot responds with `"hi there\nI'm here~But not there"` the backend will parse that output as `["hi there", "I'm here", "But not there"]`.

!!! warning
    This setting must be in sync with the questions excel and the chatbot's questions.

- `bot_question_message_flag`: string to define the messages trigerred by the notification. It will just echo the input message. Defaults to `<NOTIFICATION_QUESTION>`.

## Testing

- `username_test_user`: test user's username. It's only used during tests. Defaults to `the_Test`.
- `username_test_password`: test user's password. It's only used during tests. Defaults to `the_TestPassword`.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on <a href="https://keepachangelog.com/en/1.0.0/" class="external-link" target="_blank">Keep a Changelog</a>,
and this project adheres to <a href="https://semver.org/spec/v2.0.0.html" class="external-link" taget="_blank">Semantic Versioning</a>.

## [Unreleased]

### Added

- Endpoints to create and retrieve multiples files at once (`/files/multi`).
- Endpoint to list the files grouped by language (`/files/all`)
- A `POST`, `PUT` or `DELETE` request to `/files` will trigger the image outdated detection (if an image is not referenced in any file, it will be automatically removed).

### Changed

- *Interactive* docs are now in `/idocs` (previously in `/docs`) and *redoc* docs are now in `/docs` (previously in `/redoc`).
- Disable only *interactive* docs (`/idocs`) in **production**.

### Fixed

- Fix `codecov`'s flags.

### Removed

- Disable logs rolling.
- Settings `max_logs`.

## [1.0.0] - 2021-03-04

### Fixed

- Fixed [#70](https://github.com/BelinguoAG/full-power-backend/issues/70)
- Fixed [#69](https://github.com/BelinguoAG/full-power-backend/issues/69)

## [0.13.0] - 2021-02-25

### Added

- Add endpoint `/notifications-content/second-survey/{problem}`.
- Add endpoint `/notifications-content/generic`.
- Add setting `bot_question_message_flag`.
- Add support to `python 3.9`.

### Changed

- Add parameter `question_response` to endpoint `/process-msg`.
- Enable second survey questions (echo).
- Enable second survey responses.
- Send header `X-Problems-Parsed` in the last step of `/process-msg` endpoint.
- Send header `X-Health-Data-Result` in the last step of `/process-msg` endpoint.

## [0.12.0] - 2021-02-14

### Changed

- Changed Health Data algorithm. Now considers two classes of real problems: `light` and `serious`.
- Show user's problem report in a new message in the chat.

### Fixed

- Error message in `File.name` validation.

### Removed

- Removed `settings.problem_ratio_threshold`

## [0.11.1] - 2021-02-12

### Fixed

- Script `release-changelog` fails creating a tag.

## [0.11.0] - 2021-02-12

### Added

- Add docs page.
- Send the contents of `/me` in the HTTP header `X-Current-User` as json in `/login`.
- Add attribute to **User** model: `accepted_disclaimer`
- Add attribute to **User** model: `filled_survey`
- Add endpoint to set **User.accepted_disclaimer**: `/accept-disclaimer`
- Add endpoint to set **User.filled_survey**: `/survey-filled`

### Changed

- Script `update-questions` needs an excel file, not a csv one.
- Improve scripts using `typer`.

### Fixed

- `/me` should not return the `user_id`.

## [0.10.0] - 2021-01-30

### Added

- Add attribute `user.last_login`.
- Add script `create-scripts-docs`.

### Changed

- Bump `fastapi` version to `0.63.0`.
- Improve `requirements` management.
- Endpoint `/bot/process-msg` returns a list of the bot's messages in the `bot_msg` attribute.
- **Internal** refactoring.

## [0.9.0] - 2020-12-22

### Added

- Add new setting: `problem_ratio_threshold`

### Changed

- When processing a complete HealthData, now it identifies real problems using the setting `problem_ratio_threshold` and explains them in the returned string.
- Return `409` instead of `400` in case of id conflict creating models.
- Make `POST /image` return `415` if image processing fails. ([#38](https://github.com/BelinguoAG/full-power-backend/issues/38))
- Return `201` instead of `200` on model creation.
- Return `404` in endpoints which returns a user attribute if the user does not exist.

## [0.8.1] - 2020-12-14

### Fixed

- `display_type` not showing in `POST /process/process-msg`. ([#35](https://github.com/BelinguoAG/full-power-backend/issues/35))

## [0.8.0] - 2020-12-07

### Added

- CRUD routes for managing files.
- CRUD routes for managing images.
- Cover every schema with tests.

### Changed

- Improve documentation.
- Improve CRUD pattern.
- Improve schemas.
- Remove testing database before each unit test.
- Pin requirements versions

### Fixed

- Check for username validity before registering a new user.

## [0.7.1] - 2020-11-30

### Fixed

- Remove "v" from version name (#27).

## [0.7.0] - 2020-11-29

### Added

- Set up database migrations.

## [0.6.0] - 2020-11-27

### Added

- Setup github actions (CI) and codecov as coverage analyser.
- Add badges to readme.

### Changed

- Refactor entire application based on [full-stack-fastapi-postgresql](https://github.com/tiangolo/full-stack-fastapi-postgresql/tree/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app)
- Improve bot manager

### Fixed

- Enable CI and coverage analyser

### Fixed

- Rollback database if error detected.

## [0.5.1] - 2020-10-24

### Fixed

- Fix permission error when doing a log rollover.

## [0.5.0] - 2020-10-22

### Added

- Add route `/refresh` to refresh tokens.
- Add scopes to tokens.
- Add setting `TOKEN_EXPIRE_MINUTES`.

## [0.4.1] - 2020-10-21

### Fixed

- Fix route DELETE `/users/<id>`.

## [0.4.0] - 2020-10-21

### Added

- Add `LOG_PATH`, `MAX_LOGS` and `LOGGING_LEVEL` settings.
- Add routes `/` and `/version`.
- Enable logging.
- Log critical errors.

### Changed

- Improve login error messages via `X-Error-Reason` header.
- Move app declaration to `app.main`.
- Return token expiration time in `/login` with the token.

## [0.3.1] - 2020-10-21

### Fixed

- Specify `/settings` return model.

## [0.3.0] - 2020-10-20

### Added

- 🔥 Add WSGI conversion.
- Add `/settings` route to retrieve settings.

### Changed

- Manage `dialogflow` data via settings.

### Fixed

- Specify table's fields length to be compatible with sql.

## [0.2.0] - 2020-10-18

### Changed

- Use `pydantic`'s `BaseSettings` to manage settings.

## [0.1.0] - 2020-10-18

### Added

- First version

[unreleased]: https://github.com/BelinguoAG/full-power-backend/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/BelinguoAG/full-power-backend/compare/v0.13.0...v1.0.0

app/core/config.py
"""Internal settings of the API."""

import logging
import os
import secrets
from enum import Enum
from pathlib import Path

from pydantic import BaseSettings, FilePath, confloat


class ValidLoggingLevel(Enum):
    DEBUG = "DEBUG"
    INFO = "INFO"
    WARNING = "WARNING"
    ERROR = "ERROR"
    CRITICAL = "CRITICAL"

    def as_python_logging(self):
        return logging._nameToLevel[self.value]


class Settings(BaseSettings):
    """Internal settings of the API"""

    # General
    production: bool = False

    # Security
    encryption_algorithm: str = "HS256"
    server_secret: str = secrets.token_urlsafe(32)
    token_expire_minutes: int = 30

    # Database
    first_superuser_password: str
    first_superuser: str
    sqlalchemy_database_url: str

    # Dialogflow
    dialogflow_project_id: str
    google_application_credentials: FilePath

    # Logging
    log_path: Path
    logging_level: ValidLoggingLevel = ValidLoggingLevel.INFO

    # Bot generics
    bot_split_char: str = "~"
    bot_question_message_flag: str = "<NOTIFICATION_QUESTION>"

    # Testing
    username_test_user: str = "the_Test"
    username_test_password: str = "the_TestPassword"

    class Config:
        env_file = Path(__file__).parent.parent.with_name(".env").as_posix()
        env_file_encoding = "utf-8"

    def set_environment(self):
        os.environ[
            "GOOGLE_APPLICATION_CREDENTIALS"
        ] = self.google_application_credentials.as_posix()


settings = Settings()
settings.set_environment()
